staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go

```diff
@@ -552,16 +652,45 @@ func (alloc *allocator) allocateOne(r deviceIndices) (bool, error) {
 		return alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex + 1})
 	}
 
+	requestKey := requestIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: r.subRequest, subRequestIndex: r.subRequestIndex}
+	requestData := alloc.requestData[requestKey]
+	// If the request has subRequests, attempt them in order until one can be satisfied
+	// or all have been attempted.
+	if requestData.hasSubRequests {
+		for i := 0; ; i++ {
+			_, ok := alloc.requestData[requestIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: true, subRequestIndex: i}]
+			if !ok {
+				return false, nil
+			}
+			success, err := alloc.allocateOne(deviceIndices{claimIndex: r.claimIndex, requestIndex: r.requestIndex, subRequest: true, subRequestIndex: i})
+			if err != nil {
+				return false, err
+			}
+			if success {
+				// Store the index of the selected subRequest
+				requestData.selectedSubRequestIndex = i
+				alloc.requestData[requestKey] = requestData
+				return true, nil
+			}
+		}
+	}
```

Comment:
I think a `return` is missing after the for loop here - see https://github.com/mortent/kubernetes/pull/1/files#r1883896302

We need a test case which covers this. I haven't thought about what that test case might have to look like yet.

Reply:
The for loop doesn't have a condition (https://github.com/mortent/kubernetes/blob/DRAPrioritizedList/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L696), so it will never exit through other means than the `return` calls within the loop. If we have attempted all the subrequests, it will exit here: https://github.com/mortent/kubernetes/blob/DRAPrioritizedList/staging/src/k8s.io/dynamic-resource-allocation/structured/allocator.go#L699-L702. So a return statement will be unreachable and unnecessary. I've added a comment about this to the PR since it is not obvious when reading the code.